commit dae42a02db429d6fe809250d35b1218b9f775bcf
Author: nate zhou <gnuunixchad@outlook.com>
Date:   Fri Jan 30 13:05:12 2026 +0800

    Feat(bar): visible windows count in monocle layout
    
    This mimics the default dwm's monocle layout indicator, user can see how
    many windows are stacked in the monocle layout without having to cycle
    through all of them.
    
    - Behaviors
        + When there's no non-floating visible windows on the output,
          display the indicator defined in `config.zig`(`[=]` by default).
        + Else replace the indicator with `[n]`, where `n` is the
          non-floating visible windows count.
    
    (Sticky windows are counted if they are non-floating.)

diff --git a/src/kwm/bar.zig b/src/kwm/bar.zig
index 1c01fb8..9428115 100644
--- a/src/kwm/bar.zig
+++ b/src/kwm/bar.zig
@@ -499,13 +499,42 @@ fn render_dynamic_component(self: *Self) void {
     var x: i16 = 0;
     const y: i16 = 0;
 
-    x += self.render_str(
-        buffer,
-        config.layout_tag(self.output.current_layout()),
-        &normal_fg,
-        x+@as(i16, @intCast(@divFloor(pad, 2))),
-        y,
-    ) + @as(i16, @intCast(pad));
+    const current_layout = self.output.current_layout();
+    if (current_layout == .monocle) {
+        var visible_count: usize = 0;
+        {
+            var it = context.windows.safeIterator(.forward);
+            while (it.next()) |window| {
+                if (window.is_visible_in(self.output) and !window.floating) {
+                    visible_count += 1;
+                }
+            }
+        }
+
+        const layout_symbol = if (visible_count == 0)
+            config.layout_tag(.monocle)
+        else blk: {
+            var buf: [16]u8 = undefined;
+            const formatted = std.fmt.bufPrint(&buf, "[{}]", .{visible_count}) catch "[?]";
+            break :blk formatted;
+        };
+
+        x += self.render_str(
+            buffer,
+            layout_symbol,
+            &normal_fg,
+            x+@as(i16, @intCast(@divFloor(pad, 2))),
+            y,
+        ) + @as(i16, @intCast(pad));
+    } else {
+        x += self.render_str(
+            buffer,
+            config.layout_tag(current_layout),
+            &normal_fg,
+            x+@as(i16, @intCast(@divFloor(pad, 2))),
+            y,
+        ) + @as(i16, @intCast(pad));
+    }
 
     const mode_tag =
         if (config.mode_tag.contains(context.mode)) config.mode_tag.getAssertContains(context.mode)
